xds/src/main/java/io/grpc/xds/EdsLoadBalancer.java

```diff
@@ -0,0 +1,737 @@
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "EDS load balancer is shutting down");
+
+    if (endpointWatcher != null) {
+      xdsClient.cancelEndpointDataWatch(edsServiceName, endpointWatcher);
+      endpointWatcher.reset();
+    }
+  }
+
+  static final class EdsConfig {
+
+    /**
+     * Name of cluster to query EDS for.
+     */
+    final String name;
+
+    final Object localityPickingPolicy;
+
+    EdsConfig(String name, Object localityPickingPolicy) {
+      this.name = checkNotNull(name, "name");
+      this.localityPickingPolicy = checkNotNull(localityPickingPolicy);
+    }
+  }
+
+  static final class EndpointWatcherImpl implements EndpointWatcher {
+
+    private static final String ROUND_ROBIN = "round_robin";
+    private static final long DELAYED_DELETION_TIMEOUT_MINUTES = 15L;
+    private static final EndpointWatcherImpl.PickerFactory pickerFactoryImpl =
+        new EndpointWatcherImpl.PickerFactory() {
+          @Override
+          public SubchannelPicker picker(List<WeightedChildPicker> childPickers) {
+            return new InterLocalityPicker(childPickers);
+          }
+        };
+
+    private final Helper helper;
+    private final PickerFactory pickerFactory;
+    private final LoadBalancerProvider loadBalancerProvider;
+    private final ThreadSafeRandom random;
+    private final LoadStatsStore loadStatsStore;
+    private final OrcaPerRequestUtil orcaPerRequestUtil;
+    private final OrcaOobUtil orcaOobUtil;
+    private final PriorityManager priorityManager = new PriorityManager();
+    private final Map<Locality, LocalityLbInfo> localityMap = new HashMap<>();
+    // Most current set of localities instructed by traffic director
+    private Set<Locality> localities = ImmutableSet.of();
+    private List<DropOverload> dropOverloads = ImmutableList.of();
+    private long metricsReportIntervalNano = -1;
+
+    EndpointWatcherImpl(
+        Helper helper,
+        PickerFactory pickerFactory,
+        LoadBalancerRegistry lbRegistry,
+        ThreadSafeRandom random,
+        LoadStatsStore loadStatsStore,
+        OrcaPerRequestUtil orcaPerRequestUtil,
+        OrcaOobUtil orcaOobUtil) {
+      this.helper = checkNotNull(helper, "helper");
+      this.pickerFactory = checkNotNull(pickerFactory, "pickerFactory");
+      loadBalancerProvider = checkNotNull(
+          lbRegistry.getProvider(ROUND_ROBIN),
+          "Unable to find '%s' LoadBalancer", ROUND_ROBIN);
+      this.random = checkNotNull(random, "random");
+      this.loadStatsStore = checkNotNull(loadStatsStore, "loadStatsStore");
+      this.orcaPerRequestUtil = checkNotNull(orcaPerRequestUtil, "orcaPerRequestUtil");
+      this.orcaOobUtil = checkNotNull(orcaOobUtil, "orcaOobUtil");
+    }
+
+    @Override
+    public void onEndpointChanged(EndpointUpdate update) {
+      checkNotNull(update, "update");
```

Comment:
Useless check, remove.

Reply:
Although not working on this now. What is your criteria of useless/useful check? Comparing with https://github.com/grpc/grpc-java/pull/6320/files#diff-2ab746208e54edd7bcbba5fc408b6946R149

Reply:
Hmm... It's hard to tell the exact criteria for when to put a check. Sometimes I may not consider too much for adding a check, just feel like "this argument comes from someone other than the implementor, he/she may give me something unexpected". Seems isn't that objective and I may revisit and remove some unnecessary checks.

But in general, I have a sense of some conditions that a check is good to have: If something is wrong but not broken immediately.

For example, if a method/constructor takes in some argument and assigns it to a field of the class. That is, the passed-in argument becomes part of the object state and is saved to be used later. We want a checkNotNull for it as we want to know the first place of being wrong instead of the place it is used and caused breakage. In my case, `configWatcher` is saved and used later, so we would want it break at its earliest place instead of some time later. For your case, you are not saving `update` for later uses, if it is null, it will break immediately as you call `update. getDropPolicies()` and we could know immediately what went wrong.

For `checkState`, it's more complicated and may involve more throughout considerations.